### PR TITLE
Api 27812 bdd claim type validation

### PR DESCRIPTION
**Pull request by mchristiansonVA**
## Summary
- If claimProcessType is "BDD_PROGRAM", form submission must have an activeDutyEndDate or anticipatedSeparationDate between 90 & 180 days from claimDate.

## Related issue(s)
[API-27812](https://jira.devops.va.gov/browse/API-27812)

## Testing done

- Local testing w/ Postman
Set claimProcessType to "BDD_PROGRAM"
Set activeDutyEndDate or anticipatedSeparationDate to value between 90 & 180 days from claimDate; confirm no validation error.
Set activeDutyEndDate or anticipatedSeparationDate to value outside of range of 90 & 180 days from claimDate; confirm validation error w/ message "Must have an activeDutyEndDate or anticipatedSeparationDate between 90 & 180 days from claim date."

Note that 526 submissions with claimProcessType set to "BDD_PROGRAM" that pass validation will have the claimProcessType value mapped to "BDD_PROGRAM_CLAIM" for submission to the EVSS docker container and PDF Generator services. For testing I confirmed there were no EVSS errors in the console and that it was submitted successfully to EVSS, as well as confirming that the generated 21-526EZ pdf has the correct checkbox checked in Section 1 of the form for program type "BDD Program Claim".

- Rspec tests added for validation on 526 submit as well as mapping for EVSS and PDF Generator.

## What areas of the site does it impact?
modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb
modules/claims_api/lib/claims_api/v2/disability_compensation_evss_mapper.rb
modules/claims_api/lib/claims_api/v2/disability_compensation_pdf_mapper.rb
modules/claims_api/spec/requests/v2/veterans/disability_compensation_request_spec.rb
modules/claims_api/spec/lib/claims_api/v2/disability_compensation_evss_mapper_spec.rb
modules/claims_api/spec/lib/claims_api/v2/disability_compensation_pdf_mapper_spec.rb

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 1]
```diff
@@ -833,8 +834,51 @@ def raise_exception_if_value_not_present(val, form_obj_description)
         )
       end
 
+      def validate_claim_process_type_bdd!
+        claim_process_type = form_attributes['claimProcessType']
+        if claim_process_type == 'BDD_PROGRAM'
```
Comment (stiehlrod) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:839]: the claim_type is wrong, it should be "BDD_PROGRAM_CLAIM"
 ```Unmapped VA900 (Backend Response: { status: 415, message: {"errors":[{"id":"3f242539-51fd-4f72-9a1f-301b17502836","status":"415","code":"HttpMessageNotReadable","title":"Unsupported Media Type","detail":"JSON parse error: Cannot deserialize value of type `gov.va.api.lighthouse.benefits.form526.api.v1.ClaimProcessTypeV1` from String \"BDD_PROGRAM\": not one of the values accepted for Enum class: [IDES, FDC_PROGRAM, BDD_PROGRAM_CLAIM, STANDARD_CLAIM_PROCESS]","meta":{"time":"2023-10-...```

Comment (mchristiansonVA) [reply modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:839]: Discussed with Kayla: Form 526 submissions will have a value of "BDD_PROGRAM" as appropriate for the claimant, but we now map these values to "BDD_PROGRAM_CLAIM" for submission to the EVSS docker container and the PDF Generator.

**Review by rockwellwindsor-va — COMMENTED** [review]
I was testing this out and just had a question.  I created two claims while testing this and then looked then up in VBMS and neither had the BDD Program box checked but I wanted to verify I am looking at the right spot/not missing anything.  With "claimProcessType": "BDD_PROGRAM", set on the claim should the BDD box be checked on these? (Shared images in the PR channel for this PR)

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 2]
```diff
@@ -833,8 +834,51 @@ def raise_exception_if_value_not_present(val, form_obj_description)
...
+          date = form_attributes['claimDate'] || Time.find_zone!('Central Time (US & Canada)').today
```
Comment (FonzMP) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:840]: for curiosity, what is the decision for Central Time here? not sure I see anything in the ticket so just getting clarity.

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 3]
```diff
@@ -833,8 +834,51 @@ def raise_exception_if_value_not_present(val, form_obj_description)
...
+          claim_date = Date.parse(date.to_s)
```
Comment (FonzMP) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:841]: should we just wrap `form_attributes['claimDate']` in a Date.parse above? 

form_attributes is a string that we cast to a string to return a date and the `Time.find_zone` is a date we cast to a string that gets returned as a date in this line. we're duplicating the process somewhere, I feel like `Date.parse(form_attributes['claim_date'])` may make more sense?

```
(byebug) claim_date
Fri, 18 Dec 2020
(byebug) claim_date.class
Date
(byebug) Time.find_zone!('Central Time (US & Canada)').today
Fri, 20 Oct 2023
(byebug) Time.find_zone!('Central Time (US & Canada)').today.class
Date
```

whereas you could do
```
claim_date =  Date.parse(form_attributes['claimDate']) || Time.find_zone!('Central Time (US & Canada)').today

(byebug) claim_date
Fri, 18 Dec 2020
```

seems to function the same. with my example I'm seeing
```
{
    "errors": [
        {
            "title": "Unprocessable Entity",
            "detail": "Must have an activeDutyEndDate or anticipatedSeparationDate between 90 & 180 days from claim date."
        }
    ]
}
```

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 4]
```diff
@@ -833,8 +834,51 @@ def raise_exception_if_value_not_present(val, form_obj_description)
...
+          date_in_bdd_claim_date_range(Date.strptime(anticipated_seperation_date, '%m-%d-%Y'), claim_date)
+        else
+          false
+        end
+      end
+
+      def date_in_bdd_claim_date_range(target_date, claim_date)
+        target_date.between?(claim_date.next_day(90), claim_date.next_day(180))
+      end
+
```
Comment (FonzMP) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:881]: I think the entirety of changes in this file could roughly translate to something a bit simpler.
```
      def validate_claim_process_type_bdd!
        claim_process_type = form_attributes['claimProcessType']
        if claim_process_type == 'BDD_PROGRAM'
          claim_date = Date.parse(form_attributes['claimDate']) || Time.find_zone!('Central Time (US & Canada)').today
          service_information = form_attributes['serviceInformation']
          active_dates = service_information['servicePeriods']&.pluck('activeDutyEndDate')
          active_dates << service_information&.dig('federalActivation', 'anticipatedSeparationDate')

          unless active_dates.compact.map do |a|
                   Date.strptime(a, '%m-%d-%Y').between?(claim_date.next_day(90), claim_date.next_day(180))
                 end.any?
            raise ::Common::Exceptions::UnprocessableEntity.new(
              detail: 'Must have an activeDutyEndDate or anticipatedSeparationDate between 90 & 180 days' \
                      ' from claim date.'
            )
          end
        end
      end
```

but this is just a rough take at it with the example I have. I have not thoroughly tested. I feel like we can get away with moving this to just plucking the dates, iterating through and checking for any between the required dates. Seems like we're passing specific arrays in and have multiple functions to handle specific formats where we could just pass dates to one consistent method. 

If we want to go real crazy we could write constants for 90 and 180 so they're easy to change. If we update the constant we get it in the `next_day(BDD_LOWER_LIMIT)` and the error with something like `between #{BDD_UPPER_LIMIT}`.

The date manipulation in the unless line I think can be removed when we go back to (wait for it...... ) ISO dates. Sorry, I had to 😆 



I would almost take it one step further and add a method for the first line and the if, and only call line 35 if it's true.. so something like  `validate_claim_process_type_bdd! if bdd_claim?`

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb (hunk 2 above)
Comment (mchristiansonVA) [reply modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:840]: See validate_form_526_submission_claim_date comment, line 105 in the same file:
 # EVSS runs in the Central US Time Zone.
 # So 'claim_date' needs to be <= current day according to the Central US Time Zone.

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb (hunk 3 above)
Comment (mchristiansonVA) [reply modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:841]: I duplicated the logic for handling the claim date from validate_form_526_submission_claim_date! in line 105 of this file. I agree the existing approach could use some improvement...I'll review and make updates.

Comment (mchristiansonVA) [reply modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:841]: The existing approach handles the case when claimDate is null, so leaving as-is for now.

**Pull request closed without merging**